tests/test_device_tracker.py

```diff
@@ -0,0 +1,204 @@
+        tracker.history = {"a": [1]}
+        tracker.latest = {"a": 1}
+
+        tracker.reset()
+
+        assert tracker.totals == dict()
+        assert tracker.history == dict()
+        assert tracker.latest == dict()
+
+    def test_enter_and_exit(self):
+        """Assert entering and exit work as expected"""
+
+        tracker = Tracker()
+        tracker.totals = {"a": 1}
+        tracker.history = {"a": [1]}
+        tracker.latest = {"a": 1}
+
+        returned = tracker.__enter__()
+
+        assert id(tracker) == id(returned)
+        assert tracker.tracking == True
+
+        assert tracker.totals == dict()
+        assert tracker.history == dict()
+        assert tracker.latest == dict()
+
+        tracker.__exit__(1, 1, 1)
+
+        assert tracker.tracking == False
+
+    def test_context(self):
+        """Assert works with runtime context"""
+
+        with Tracker() as tracker:
+            assert isinstance(tracker, Tracker)
+            assert tracker.tracking == True
+
+        assert tracker.tracking == False
+
+    def test_update(self):
+        """Checks update stores to history and totals"""
+
+        tracker = Tracker()
+
+        tracker.update(a=1, b="b", c=None)
+        tracker.update(a=2, b="b2", c=1)
+
+        assert tracker.history == {"a": [1, 2], "b": ["b", "b2"], "c": [None, 1]}
+
+        assert tracker.totals == {"a": 3, "c": 1}
+
+        assert tracker.latest == {"a": 2, "b": "b2", "c": 1}
+
+    def test_record_callback(self, mocker):
+        class callback_wrapper:
+            @staticmethod
+            def callback(totals, history, latest):
+                pass
+
+        wrapper = callback_wrapper()
+        spy = mocker.spy(wrapper, "callback")
+
+        tracker = Tracker(callback=wrapper.callback)
+
+        tracker.totals = {"a": 1, "b": 2}
+        tracker.history = {"a": [1], "b": [1, 1]}
+        tracker.latest = {"a": 1, "b": 1}
+
+        tracker.record()
+
+        _, kwargs_called = spy.call_args_list[-1]
+
+        assert kwargs_called["totals"] == tracker.totals
+        assert kwargs_called["history"] == tracker.history
+        assert kwargs_called["latest"] == tracker.latest
+
+
+class TestDefaultTrackerIntegration:
+    def test_single_execution_default(self, mocker):
+        """Test correct behavior with single circuit execution"""
+
+        class callback_wrapper:
+            @staticmethod
+            def callback(totals=dict(), history=dict(), latest=dict()):
+                pass
+
+        wrapper = callback_wrapper()
+        spy = mocker.spy(wrapper, "callback")
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit():
+            return qml.expval(qml.PauliZ(0))
+
+        with Tracker(circuit.device, callback=wrapper.callback) as tracker:
+            circuit()
+
+        assert tracker.totals == {"executions": 1}
+        assert tracker.history == {"executions": [1], "shots": [None]}
```

Comment:
Would be great to test calling `circuit` multiple times when `shots=None`. Maybe parametrizing this test case and the next one to have:
1. 1 circuit execution
2. 2 circuit exeuctions

could help to cover that the totals and the history values are good.